### DISCUSSION TITLE
Draw line through dots

### DISCUSSION THREAD

**Discussion by erwindenboer** [ManimCommunity/manim#1800] [General] [answered]
Hello,

For differentiation, I want to draw a line which goes through two dots. For now, I can draw a line which ends at two dots. Is there something possible which does this?
![image](https://user-images.githubusercontent.com/31764383/125797384-7383e4d2-6b2b-4f90-ba85-ee4f6859813e.png)

**Comment by kilacoda-old** [comment, marked as answer]
You could try to do what you're doing and add multiples of `LEFT` and `RIGHT` to each endpoint of the line:

```py
Line(start=p1 +LEFT, end=p2+RIGHT)
```

EDIT: See last few replies to this thread for the real solution.

**Reply by erwindenboer** [reply]
@kilacoda Thanks for your fast response. I forgot to mention that I use animations. I want to explain how differentiation works. For this, I want to move two points closer together. If I integrate LEFT and RIGHT this becomes static. Is there a dynamic way to move the line? My code is included.
[code.txt](https://github.com/ManimCommunity/manim/files/6823607/code.txt)

**Reply by kilacoda-old** [reply]
You could get the vector between the two points and subtract/add multiples of that:
```py

        def vector_between_points(dot1: Dot, dot2: Dot):
            return dot2.get_center() - dot1.get_center()

        line_0.add_updater(
            lambda z: z.put_start_and_end_on(
                    dot_0.get_center() - vector_between_points(dot_0, dot_1),
                    dot_1.get_center() + vector_between_points(dot_0, dot_1)
            )
        )
```

**Reply by erwindenboer** [reply]
Unfortunately, this is also not what I want. For now, I get one line which stops at the dots. The problem is that I want them to go through, as can be seen in the picture.

![image](https://user-images.githubusercontent.com/31764383/125841026-3cc9dfb1-ef53-499c-bee5-74cfa9658d7e.png)

The same way as the red and red dotted lines.

**Reply by hydrobeam** [reply]
[`get_secant_slope_group`](https://docs.manim.community/en/stable/reference/manim.mobject.coordinate_systems.CoordinateSystem.html#manim.mobject.coordinate_systems.CoordinateSystem.get_secant_slope_group) seems like it meets your needs.

```py
class slope(Scene):
    def construct(self):
        ax = Axes()
        gr = ax.get_graph(lambda x : x**2)
        slopes = ax.get_secant_slope_group(x=1.2, graph=gr, dx=0.4, dx_label="x", dy_label="y")

        self.add(ax, gr, slopes) 
```
![image](https://user-images.githubusercontent.com/70682032/125842269-cf4a9949-20de-4dcc-9ba1-70914dcdd9d8.png)

**Reply by erwindenboer** [reply]
@hydrobeam Thank you! That's a big leap to what I want. However, I get the triangle labeled x and y which I don't want. I only want the line to cross the two points. Besides, the line is only the secant and can't cross the two points. Is there another way?

![image](https://user-images.githubusercontent.com/31764383/125859189-b3aae5f5-166b-4d90-819c-87f6c69aebf9.png)

The above image reflects what I have in mind. You can see two lines, one is labeled m and the other isn't labeled but crosses the points P and Q. I want the latter line to transform to line m. I want this to be done by moving point Q towards point P so that line m is found. It's the transform which makes it "click" for the students in my opinion.

**Reply by erwindenboer** [reply]
Is there someone who has an idea? @kilacoda @hydrobeam

**Reply by kilacoda-old** [reply]
@erwindenboer Something like this?

https://user-images.githubusercontent.com/65204531/127093452-f97d72e0-6f0d-4a22-92d0-dde0d614f439.mp4

I just applied `.scale(4)` to `line_0` in its definition and the updater

**Reply by erwindenboer** [reply]
@kilacoda That's exactly what I mean. Thank you! Is it also possible to let the line stop at the borders of the screen? Otherwise, I need to use a large number (e.g. .scale(90000)). Unfortunately, I can't mark this reaction as an answer. But I'll mark your first answer as answered.

**Reply by erwindenboer** [reply]
@kilacoda Can you send your code on how you created this?

**Reply by kilacoda-old** [reply]
> @kilacoda That's exactly what I mean. Thank you! Is it also possible to let the line stop at the borders of the screen? Otherwise, I need to use a large number (e.g. .scale(90000)). Unfortunately, I can't mark this reaction as an answer. But I'll mark your first answer as answered.

You probably don't need *that* big of a scale factor 😅 , but something in the range of 25-50 should be more than enough.



> @kilacoda Can you send your code on how you created this?

Sure, though it's a bit modified from your original `manim_presentation` based code:

```py
from manim import *

class Begin(Scene):
    def construct(self):
        ax = Axes(
            x_range=[-2, 5, 1],
            y_range=[-5, 5, 1],
            x_length=12,
            y_length=7,
            x_axis_config={"numbers_to_include": np.arange(-2, 6, 1)},
            y_axis_config={"numbers_to_include": np.arange(-5, 6, 1)},
            tips=False
        )

        labels = ax.get_axis_labels()

        f1 = lambda x: x**3-4*x**2+3*x+1

        func_1 = ax.get_graph(f1, color = ORANGE)

        t_0 = ValueTracker(0)
        t_1 = ValueTracker(1)

        dot_0 = Dot(point=ax.coords_to_point(t_0.get_value(), f1(t_0.get_value())), color=PURPLE)
        dot_1 = Dot(point=ax.coords_to_point(t_1.get_value(), f1(t_1.get_value())), color=PURPLE)

        dot_0.add_updater(lambda x: x.move_to(ax.c2p(t_0.get_value(), f1(t_0.get_value()))))
        dot_1.add_updater(lambda x: x.move_to(ax.c2p(t_1.get_value(), f1(t_1.get_value()))))

        line_0 = Line(dot_0.get_center(),dot_1.get_center(), color=PURPLE).scale(4)
        line_0.add_updater(lambda z: z.become(Line(dot_0.get_center(), dot_1.get_center()).scale(4)))

        self.play(FadeIn(ax, func_1, dot_0, dot_1, line_0))
        self.play(t_1.animate.set_value(0.5))
        self.wait()
```
